Sources/MapboxNavigation/NavigationViewHelper.swift

```diff
@@ -0,0 +1,170 @@
+import CoreLocation
+import UIKit
+
+
+class NavigationViewHelper: NavigationComponent, ViewCycleObserver {
+    
+    // MARK: - Properties
+    
+    weak var navigationViewData: NavigationViewData!
+    var annotatesSpokenInstructions = false
+    
+    fileprivate var navigationMapView: NavigationMapView {
+        return navigationViewData.navigationView.navigationMapView
+    }
+    
+    fileprivate var router: Router! {
+        navigationViewData.navigationService.router
+    }
+    
+    fileprivate var routeLineTracksTraversal: Bool {
+        get {
+            navigationMapView.routeLineTracksTraversal 
+        }
+        set {
+            navigationMapView.routeLineTracksTraversal = newValue
+        }
+    }
+    
+    private var currentLegIndexMapped = 0
+    private var currentStepIndexMapped = 0
+    
+    // MARK: - Public methods
+    
+    func populateName(for waypoint: Waypoint, populated: @escaping (Waypoint) -> Void) {
+        guard waypoint.name == nil else { return populated(waypoint) }
+        let location = CLLocation(latitude: waypoint.coordinate.latitude, longitude: waypoint.coordinate.longitude)
+        CLGeocoder().reverseGeocodeLocation(location) { (places, error) in
+        guard let place = places?.first, let placeName = place.name, error == nil else { return }
+            let named = Waypoint(coordinate: waypoint.coordinate, name: placeName)
+            return populated(named)
+        }
+    }
+    
+    // MARK: - Private methods
+    
+    fileprivate func showRouteIfNeeded() {
+        guard navigationViewData.navigationViewController.isViewLoaded &&
+                navigationViewData.navigationViewController.view.window != nil else { return }
+        guard !navigationMapView.showsRoute else { return }
+        navigationMapView.show([router.route], legIndex: router.routeProgress.legIndex)
+        navigationMapView.showWaypoints(on: router.route, legIndex: router.routeProgress.legIndex)
+        
+        let currentLegProgress = router.routeProgress.currentLegProgress
+        let nextStepIndex = currentLegProgress.stepIndex + 1
+        
+        if nextStepIndex <= currentLegProgress.leg.steps.count {
+            navigationMapView.addArrow(route: router.route, legIndex: router.routeProgress.legIndex, stepIndex: nextStepIndex)
+        }
+
+        if annotatesSpokenInstructions {
+            navigationMapView.showVoiceInstructionsOnMap(route: router.route)
+        }
+    }
+    
+    fileprivate func makeGestureRecognizersResetFrameRate() {
+        for gestureRecognizer in navigationMapView.mapView.gestureRecognizers ?? [] {
+            gestureRecognizer.addTarget(self, action: #selector(resetFrameRate(_:)))
+        }
+    }
+    
+    @objc func resetFrameRate(_ sender: UIGestureRecognizer) {
+        navigationMapView.mapView.update {
+            $0.render.preferredFramesPerSecond = NavigationMapView.FrameIntervalOptions.defaultFramesPerSecond
+        }
+    }
```

Comment:
Frame rate management appears to be entirely dependent on NavigationMapView; can we move this functionality to that class?

Reply:
Done